### ISSUE TITLE
train with multi-gpu with MirroredStrategy will hang-up

### ISSUE THREAD

**Issue by honeytidy** [tensorflow/tensorflow#22889]
### System information
Have I written custom code: N/A
OS Platform and Distribution: CentOS Linux release 7.3.1611
TensorFlow installed from:  (pip install tf-nightly-gpu)
TensorFlow version: Tensorflow('v1.9.0-rc2-5345-g57d31aa599', '1.12.0-dev20181005')
Bazel version: N/A
GPU model and memory: Tesla P40 24G
Exact command to reproduce: N/A
Mobile device: N/A
CUDA/cuDNN version: cuda 9.0 with cudnn7.1.4

I train with tensorflow for multi-gpu with MirroredStrategy and estimator. I got the problem:
when I set the distribute mode with the following code it will got stuck after runing some training steps:
```
distribution = tf.contrib.distribute.MirroredStrategy()
config = tf.estimator.RunConfig(train_distribute=distribution)
estimator = tf.estimator.Estimator(model_fn=mymodel_fn, model_dir='logs',
        config=config)
```
bug when I run without distribute mode like this:
```
distribution = tf.contrib.distribute.MirroredStrategy()
config = tf.estimator.RunConfig()
estimator = tf.estimator.Estimator(model_fn=mymodel_fn, model_dir='logs',
        config=config)
```
It runs ok. Why? 
Is that a bug of MirroredStrategy?

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
GPU model and memory
Exact command to reproduce
Mobile device

**Comment by honeytidy** [comment]
OK, question updated

**Comment by magnofel** [comment]
Have I written custom code: N/A
OS Platform and Distribution: Ubuntu 18.04
TensorFlow installed from: (pip install tensorflow-gpu)
TensorFlow version: 1.11.0
Bazel version: N/A
GPU model and memory: 1080ti
Exact command to reproduce: see code beliw
Mobile device: N/A
CUDA/cuDNN version: cuda 10.0 with cudnn7.1.4


`import tensorflow as tf
def model_fn(features, labels, mode):
  layer = tf.layers.Dense(1)
  logits = layer(features)

  if mode == tf.estimator.ModeKeys.PREDICT:
    predictions = {"logits": logits}
    return tf.estimator.EstimatorSpec(mode, predictions=predictions)

  loss = tf.losses.mean_squared_error(
      labels=labels, predictions=tf.reshape(logits, []))

  if mode == tf.estimator.ModeKeys.EVAL:
    return tf.estimator.EstimatorSpec(mode, loss=loss)

  if mode == tf.estimator.ModeKeys.TRAIN:
    train_op = tf.train.GradientDescentOptimizer(0.2).minimize(loss)
    return tf.estimator.EstimatorSpec(mode, loss=loss, train_op=train_op)

def input_fn():
  features = tf.data.Dataset.from_tensors([[1.]]).repeat(100)
  labels = tf.data.Dataset.from_tensors(1.).repeat(100)
  return tf.data.Dataset.zip((features, labels))


distribution = tf.contrib.distribute.MirroredStrategy()
config = tf.estimator.RunConfig(train_distribute=distribution)
classifier = tf.estimator.Estimator(model_fn=model_fn, config=config)
classifier.train(input_fn=input_fn)
classifier.evaluate(input_fn=input_fn)`

**Comment by tensorflowbutler** [comment]
Nagging Assignee @seemuch: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by jnd77** [comment]
With TF 1.12 I still have the issue that @magnofel encountered (only difference since TF 1.11 is that it freezes before displaying `INFO:tensorflow:Initialize system`). 
@seemuch do you have any update on this ? Thanks a lot.

**Comment by Luonic** [comment]
Same problem as @magnofel
r1.11 (1.12 has broken distributed training)
built from source
nccl 2.3.5
@seemuch how can we debug this problem?

**Comment by Luonic** [comment]
@gunan maby you can take a look at this? Looks like the problem is at the step result aggregation. I can share all logs that needed to fix this.

**Comment by seemuch** [comment]
Haven't had bandwidth to look at this yet. 
Will take a look within this week.

**Comment by Luonic** [comment]
@seemuch i have resolved my issue and it can possibly help others here (@jnd77 @magnofel @honeytidy )
If you use AMD Treadripper and motherboard without PLX chips then you should go to UEFI and disable IOMMU. NCCL is not compatible with it. 
More you can find [here](https://github.com/NVIDIA/nccl/issues/120) and [here](https://github.com/pytorch/pytorch/issues/1637#issuecomment-338268158)

**Comment by jnd77** [comment]
Thanks a lot @Luonic. With your links, we solved the issue. 
We disabled IOMMU via grub as mentioned [here](https://github.com/pytorch/pytorch/issues/1637#issuecomment-338268158).

**Comment by Luonic** [comment]
@patzm I think cloud providers test thier instances for compatibility (at least GCP, AWS do I think). And you can always disable it with grub config as @jnd77 did.

**Comment by JesseFarebro** [comment]
This issue seems to be the most relevant to the problem I'm facing. I tried disabling NCCL P2P which should allow for training (albeit slower training) as per the issues pointed out by @Luonic. This fix doesn't seem to work. I don't have access to disable IOMMU, I was wondering if there is any other troubleshooting steps or workarounds for this issue?

For extra information, I did build from source (r1.96) using NCCL 1.3 and NCCL 2.3.5. Attempting to train on 2-4 NVIDIA P100's. I thought my issue might be related to #25057 but I left the system for an extended period of time and training never began.

**Comment by Frank1993** [comment]
I also encounter the same problem.  I used dataset to read data, the MirroredStrategy job will hang on at last batch if I use "shuffle->repeat->batch". But when I changed to "shuffle->batch->repeat", the job will finished correctly

**Comment by mohantym** [comment]
Hi @honeytidy!
It seems you are using older versions(1.x versions) of Tensorflow which is not supported any more. Please visit these links to upgrade your codebase to latest versions(2.7).Ref [1](https://www.tensorflow.org/addons),[2](https://www.tensorflow.org/guide/migrate). Thanks!
